### ISSUE TITLE
feature request: Dynamic generation of Dockerfile with CUDA base image and this toolstack

### ISSUE THREAD

**Issue by ChristophSchranz** [jupyter/docker-stacks#1015]
Hi guys,

I really enjoy this great maintained docker stack which I use for ML-projects. I've worked on replacing the base image by a GPU-ready image, here `nvidia/cuda:10.X-base-ubuntu18.04`, such that I can run my PyTorch (and Tensorflow) in Jupyter.

A tutorial was written on [medium](https://medium.com/@christoph.schranz/set-up-your-own-gpu-based-jupyterlab-e0d45fcacf43) that explains the set up. It is based on this [repo](https://github.com/iot-salzburg/gpu-jupyter) that uses nearly all of the Dockerfile from this repo. Unfortunately, that repo's Dockerfile is static and not updated whenever you improve something.

Therefore, I've created this [fork](https://github.com/ChristophSchranz/docker-stacks/tree/master/gpu-notebook) with an additional folder, in which a *GPU-based Dockerfile/Image using your toolstack* is generated dynamically. Using it, it enables all of this repo's features as well as GPU calculations and is updated by each git pull.

This means, that this directory comes with hardly any maintenance efforts, so this is an advantage in contrast to similar approaches already posted here. What do you think about it?

**Comment by maresb** [comment]
I'm not a docker-stacks maintainer, but this looks like a really cool project!  Thanks for doing this.

This makes me wonder: is there is some list of projects like this which are based off of the Jupyter Docker stacks?  That could be really useful.

By the way, I think you can greatly simplify your procedure.  Rather than concatenating all the other Dockerfiles like you're currently doing from your script, you should be able to do 
```
docker build --build-arg BASE_CONTAINER=nvidia/cuda:10.1-base-ubuntu18.04 -t gpu-notebook .
```
from your gpu-notebook directory where you just need a Dockerfile with 
```
FROM jupyter/tensorflow-notebook
```
followed by the concatenation of your `Dockerfile.pytorch` and `Dockerfile.usefulpackages` files. (I haven't tested this, but I think it should work, probably with minor modifications.)

Probably such discussions are more appropriate on your fork, but issues are disabled.  See https://softwareengineering.stackexchange.com/a/179470

Also, I've never played around with Git submodules myself, but based on my limited understanding, you may be able to include docker-stacks basically as a link from your repository, and easily update your own repository by changing the commit id of the link.

**Comment by ChristophSchranz** [comment]
Thanks @maresb !
I have though about that solution, and now tested also yours. Your solutions leads to this message:
```
$ docker build --build-arg BASE_CONTAINER=nvidia/cuda:10.1-base-ubuntu18.04 -t gpu-notebook .
Sending build context to Docker daemon  2.048kB
Step 1/1 : FROM jupyter/tensorflow-notebook
latest: Pulling from jupyter/tensorflow-notebook
5c939e3a4d10: Pulling fs layer 
...
e21a17b6313e: Pull complete 
f3adc2fcb99c: Pull complete 
Digest: sha256:e7c4416d4d7e7649068b38bfdea2eee3889b6cf4f19c487584f86fd6350f533f
Status: Downloaded newer image for jupyter/tensorflow-notebook:latest
 ---> d4e970af09be
[Warning] One or more build-args [BASE_CONTAINER] were not consumed
Successfully built d4e970af09be
Successfully tagged gpu-notebook:latest
```
and the container crashes on my system immediately after the start. I guess that the recursive pull till the BASE_IMAGE does not work as expected. Maybe, you have more luck!

By the way, I've set the issues on the fork on.

**Comment by ChristophSchranz** [comment]
Yes, the discussion should be continued in #1001

**Comment by maresb** [comment]
To answer my own question from above:

> This makes me wonder: is there is some list of projects like this which are based off of the Jupyter Docker stacks? That could be really useful.

there are the [Community Stacks](https://jupyter-docker-stacks.readthedocs.io/en/latest/using/selecting.html#community-stacks).
